File: src/main/java/net/rk4z/juef/UefOverlay.java
package net.rk4z.juef;

import static net.rk4z.juef.UefApp.logger;

public class UefOverlay implements AutoCloseable{
    private long nativeOverlayPtr;

    public UefOverlay(long nativeOverlayPtr) {
        this.nativeOverlayPtr = nativeOverlayPtr;
    }

    public void moveTo(int x, int y) {
        moveOverlay(getNativeOverlayPtr(), x, y);
    }

    public void resize(int width, int height) {
        resize(getNativeOverlayPtr(), width, height);
    }

    public void loadUrl(String url) {
        loadUrl(getNativeOverlayPtr(), url);
    }

    public void destroy() {
        destroy(getNativeOverlayPtr());
        nativeOverlayPtr = 0;
    }

//>------------------- Native methods -------------------<\\

    private native void moveOverlay(long overlayPtr, int x, int y);

    private native void resize(long overlayPtr, int width, int height);

    private native void loadUrl(long overlayPtr, String url);

    private native void destroy(long overlayPtr);

//>------------------- Native methods -------------------<\\

    /**
     * @deprecated This is a low-level API that exposes internal native pointers.
     * <p>
     * @return The native pointer of the overlay.
     */
    public long getNativeOverlayPtr() {
        return nativeOverlayPtr;
    }

    @Override
    public void close() {
        if (nativeOverlayPtr != 0) {
            logger.info("Closing overlay");
            destroy();
            nativeOverlayPtr = 0;
        }
    }
}

File: native/UefOverlay.cpp
#include "UefOverlay.hpp"
#include "UefWindow.hpp"

UefOverlay::UefOverlay(const UefWindow *window, int width, int height, int x, int y) {
    overlay_ = Overlay::Create(window->window_, width, height, x, y);
}

UefOverlay::~UefOverlay() = default;

void UefOverlay::moveTo(const int x, const int y) const {
    if (overlay_) {
        overlay_->MoveTo(x, y);
    }
}

void UefOverlay::resize(const int width, const int height) const {
    if (overlay_) {
        overlay_->Resize(width, height);
    }
}

void UefOverlay::destroy(UefOverlay* overlay) {
    delete overlay;
}

//>---------- Native methods(from java) ----------<\\

void Java_net_rk4z_uef_UefOverlay_moveTo(JNIEnv *env, jobject obj, jlong overlay_ptr, jint x, jint y) {
    auto *overlay = reinterpret_cast<UefOverlay *>(overlay_ptr);
    if (overlay) {
        overlay->moveTo(x, y);  // 安全に操作
    }
}


void Java_net_rk4z_uef_UefOverlay_destroy(JNIEnv *env, jobject obj, jlong overlay_ptr) {
    auto *overlay = reinterpret_cast<UefOverlay *>(overlay_ptr);
    if (overlay) {
        UefOverlay::destroy(overlay);
    }
}

//>---------- Native methods(from java) ----------<\\
